### PR TITLE
Bugfix: GQA with FA

### PR DESCRIPTION
**Pull request by blahBlahhhJ**
Related issue #443 

1. Added new unit tests for llama & mistral to test backward pass for the LMHead model.
2. Newer version of huggingface transformer has a new required argument for `LlamaDecoderLayer`.

Cause of the error in FA backward:
* K,V are broadcasted over Q's axis "q_heads_per_group" during forward pass but the "un-broadcast" cannot be automatically done by the FA backward kernel.

How I fix this issue:
* Since K,V are repeated over this axis, the gradient should simply be a summation over this axis. So for dK and dV, I sum over all the axis that doesn't belong to K and V (but are present in dK_j and dV_j)

### CODE REVIEW

File: src/levanter/models/flash_attention.py [hunk 1]
```diff
@@ -307,6 +307,11 @@ def do_inner_block(state):
             dV_j = dV_j + hax.dot(QPos.name, p_ij, dO_i).astype(dV_j.dtype)
             dK_j = dK_j + hax.dot(QPos.name, dAttn_ij, q_i).astype(dK_j.dtype)
 
+            # GQA-specific: eliminate unnecessary axes (e.g. 'q_heads_per_group')
```
Comment (blahBlahhhJ) [inline src/levanter/models/flash_attention.py:310]: I'm not sure if this is a good way to do this summation. If there's a better place, please let me know. I'm guessing the placement of this summation logic would affect performance. @dlwh

File: src/levanter/models/flash_attention.py [hunk 2]
```diff
@@ -307,6 +307,11 @@ def do_inner_block(state):
...
+            unnecessary_axes = hax.eliminate_axes(dV_j.axes, v.axes)
+            dV_j = hax.sum(dV_j, unnecessary_axes)
+            dK_j = hax.sum(dK_j, unnecessary_axes)
```
Comment (blahBlahhhJ) [inline src/levanter/models/flash_attention.py:313]: I'm not sure if this is a good place to do the summation. If there's a better place please let me know. I guess the placement of this logic would affect performance of FA. @dlwh

Comment (dlwh) [reply src/levanter/models/flash_attention.py:313]: wait, is this the right semantics? I think, if nothing else, you need to do this reduction before you add it to the prior dV_j/d_K_j

File: tests/test_llama.py [hunk 3]
```diff
@@ -224,6 +225,25 @@ def test_llama_lm_head_model(num_kv_heads):
...
+    input_ids = hax.random.randint(random.PRNGKey(0), (Batch, Pos), 0, Vocab.size)
+    mask = hax.nn.attention.causal_mask(Pos, llama_config.KeyPos)
+
+    llama_model = LlamaLMHeadModel.init(Vocab=Vocab, config=llama_config, key=random.PRNGKey(0))
+
+    def f(llama_model, input_ids, mask):
+        out = llama_model(input_ids, mask)
+        return hax.sum(out).scalar()
+
+    _, grads = eqx.filter_value_and_grad(f)(llama_model, input_ids, mask)
```
Comment (dlwh) [inline tests/test_llama.py:244]: can we add in test_fa that checks for all_close in the FA-vs-non-FA case for GQA? I suspect this code is incorrect

**Review by dlwh — CHANGES_REQUESTED** [review]
i think this is incorrect? let's add an explicit test in test_fa that checks equivalence of grad in the GQA case

File: tests/test_llama.py (hunk 3 above)
Comment (blahBlahhhJ) [reply tests/test_llama.py:244]: sorry I did made a mistake there.
Added a grad test in test_fa and now it passes.

File: src/levanter/models/flash_attention.py (hunk 2 above)
Comment (blahBlahhhJ) [reply src/levanter/models/flash_attention.py]: You're right. My newest revision should work. I did just randomly pick the variable names for the intermediate dot product result. let me know if you want some other meaningful variable names for them.

File: src/levanter/models/flash_attention.py [hunk 4]
```diff
@@ -304,8 +304,16 @@ def do_inner_block(state):
             dAttn_ij = p_ij * (dP_ij - D_i)
             dAttn_ij = dAttn_ij.astype(dQ_i.dtype)
 
-            dV_j = dV_j + hax.dot(QPos.name, p_ij, dO_i).astype(dV_j.dtype)
-            dK_j = dK_j + hax.dot(QPos.name, dAttn_ij, q_i).astype(dK_j.dtype)
+            pdO = hax.dot(QPos.name, p_ij, dO_i).astype(dV_j.dtype)
+            dAq = hax.dot(QPos.name, dAttn_ij, q_i).astype(dK_j.dtype)
```
Comment (dlwh) [inline src/levanter/models/flash_attention.py:308]:
```suggestion
            dV_ji = hax.dot(QPos.name, p_ij, dO_i).astype(dV_j.dtype)
            dK_ji = hax.dot(QPos.name, dAttn_ij, q_i).astype(dK_j.dtype)
```

File: src/levanter/models/flash_attention.py [hunk 5]
```diff
@@ -304,8 +304,16 @@ def do_inner_block(state):
...
+
+            # GQA-specific: eliminate unnecessary axes (e.g. 'q_heads_per_group')
+            unnecessary_axes = hax.eliminate_axes(pdO.axes, v.axes)
+            pdO = hax.sum(pdO, unnecessary_axes)
+            dAq = hax.sum(dAq, unnecessary_axes)
```
Comment (dlwh) [inline src/levanter/models/flash_attention.py:313]:
```suggestion
            dV_ji = hax.sum(dV_ji, unnecessary_axes)
            dK_ji = hax.sum(dK_ji, unnecessary_axes)
```

File: src/levanter/models/flash_attention.py [hunk 6]
```diff
@@ -304,8 +304,16 @@ def do_inner_block(state):
...
+
+            dV_j = dV_j + pdO
+            dK_j = dK_j + dAq
```
Comment (dlwh) [inline src/levanter/models/flash_attention.py:316]:
```suggestion
            dV_j = dV_j + dV_ji
            dK_j = dK_j + dK_ji
```

**Review by dlwh — APPROVED** [review]

**Comment by dlwh** [comment]
i feel like there should be something cleaner for the dot --> sum. In particular I'm pretty sure that combining the dot and the sum together will be faster.

**Pull request merged**